### PR TITLE
Commit

### PR DESCRIPTION
**Pull request by AAlchi**
### Considerations for these changes
*If applicable. Potential bugs from this change? Future additions?*

### CODE REVIEW

File: backend/src/controllers/filter-item-controller.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        const itemOwner = req.query.itemOwner as string
+        const size = req.query.size as string
+        const name = req.query.name as string
+
+        // Sort
+
+        const sortedBy = req.query.sortedBy as string
+
+        if (itemOwner && size && name) {
+            let tempData: any[] = await Item.find({})
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: Create an interface instead of using `any`

File: backend/src/controllers/filter-item-controller.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            let tempDataObject: any;
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: Create an interface instead of using `any`

File: backend/src/controllers/filter-item-controller.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            }
+
+            tempData.forEach((data) => {
+                const isIdValid = mongoose.Types.ObjectId.isValid(data._id) 
+    
+                if (!isIdValid) {
+                    return res.status(500).json({ message: "Invalid Data" }) 
+                }
+                
+                if (data.name.toLowerCase().includes(name.toLowerCase()) && data.itemOwner.toLowerCase().includes(itemOwner.toLowerCase())) {
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: `itemOwner` is a MongoDB ID, so make sure to compare it directly

File: backend/src/controllers/filter-item-controller.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                    tempDataObject = data._id
+                }  
+            })
+            fetchedInventoryItems = {itemId: tempDataObject, size: size, amount: {$gt: 1}}
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: Calling to the route should return all items with child inventoryItems that meet the above requirement, not just one that it finds

File: backend/src/controllers/filter-item-controller.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        } else if (itemOwner && !size && !name) {
+            let tempData: any[] = await Item.find({})
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: (Repeat) Create an interface instead of using `any`

File: backend/src/controllers/filter-item-controller.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            let tempDataObject: any;
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: (Repeat) Create an interface instead of using `any`

File: backend/src/controllers/filter-item-controller.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+            if (!tempData) {
+                res.status(404).json({ message: "Not Found" })
+            }
+
+            tempData.forEach((data) => { 
+                const isIdValid = mongoose.Types.ObjectId.isValid(data._id)
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: You don't have to validate `_id`, it's guaranteed to be valid

File: backend/src/controllers/filter-item-controller.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+        if (itemOwner && size && name) {
+            let tempData: any[] = await Item.find({})
+            let tempDataObject: any;
+            if (!tempData) {
+                res.status(404).json({ message: "Not Found" })
+            }
+
+            tempData.forEach((data) => {
+                const isIdValid = mongoose.Types.ObjectId.isValid(data._id)
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: Don't have to validate ID, since it's guaranteed by MongoDB to be valid. Remove this on the loops below as well

File: backend/src/controllers/filter-item-controller.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+            }
+
+            tempData.forEach((data) => {
+                const isIdValid = mongoose.Types.ObjectId.isValid(data._id) 
+    
+                if (!isIdValid) {
+                    return res.status(500).json({ message: "Invalid Data" }) 
+                }
+                
+                if (data.name.toLowerCase().includes(name.toLowerCase()) && data.itemOwner.toLowerCase().includes(itemOwner.toLowerCase())) {
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: Compare `itemOwner` with an `===` operator because it is an ID, so `itemOwner`'s ID must perfectly match the ID in `data`

File: backend/src/controllers/filter-item-controller.ts (hunk 4 above)
Comment (miguelaenlle) [reply backend/src/controllers/filter-item-controller.ts]: Issue repeats itself in the other loops too

File: backend/src/controllers/filter-item-controller.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        } else if (sortedBy == "name") {
+            let tempData: any[] = await Item.find({})
+            if (!tempData) {
+                res.status(404).json({ message: "Not Found" })
+            }
+
+            finalItems = await Item.find({}).sort({name: 1})
+        }
+
+        res.status(200).json({finalInventoryItems, finalItems})
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: The route should only return `finalItems`, filtered by `finalInventoryItems`. I'll go over this in greater detail on the call.

**Review by miguelaenlle — CHANGES_REQUESTED** [review]
Hey @AAlchi . Looks like a good start. I'll go over my feedback in greater detail on tomorrow's call, since a lot of the logic of this route is quite complicated.

File: backend/src/controllers/filter-item-controller.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                }  
+            })
+            fetchedInventoryItems += {itemId: tempDataObject, amount: {$gt: 1}}
+        }
+
+        if (!fetchedInventoryItems) {
+            res.status(404).json({ message: "Not Found" })
+        }
+
+        let finalInventoryItems = await InventoryItem.find(fetchedInventoryItems)
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts:82]: Test the edge cases -- I'm getting this error:

image

Name filter doesn't function properly

File: backend/src/routes/inventory-item-routes.ts [hunk 12]
```diff
@@ -1,5 +1,6 @@
 import { Router } from 'express'; 
 import { createInventoryItem, getIndividualInventoryItem, getInventoryItems, patchInventoryItem } from '../controllers/inventory-item-controller';
+import { loginUser } from "../controllers/auth-controller"
```
Comment (miguelaenlle) [inline backend/src/routes/inventory-item-routes.ts]: Wrong middleware -- use the `auth` middleware. Also, test creating an account, generating a token, then logging in as a user. 

I can go over this in greater detail on the call.

**Review by miguelaenlle — CHANGES_REQUESTED** [review]
Functionality overall looks good. We should cover a few more edge cases and eliminate a few more bugs before uploading this update.

I can demonstrate how to test those cases live on Monday's call, after the main portion.

File: frontend/src/Pages/Buyer/HomePageSearch.tsx [hunk 13]
```diff
@@ -0,0 +1,85 @@
+import React, { useEffect, useState } from 'react'
+
+interface HomePageSearch {
+    name: { name: string }[]
+}
+
+const data: HomePageSearch = {
```
Comment (miguelaenlle) [inline frontend/src/Pages/Buyer/HomePageSearch.tsx:7]: Component looks great, @AAlchi ! 

Dark Mode makes a lot of lighter content dark -- 
image

This can be fixed in a separate task at a later time.

File: backend/src/routes/inventory-item-routes.ts [hunk 14]
```diff
@@ -9,15 +10,15 @@ router.post('/', [
     body('size').isString().withMessage('Size must be a string').notEmpty().withMessage('Size primary is required'),
     body('price').isNumeric().withMessage('Price must be a number').notEmpty().withMessage('Price is required'),
     body('amount').isNumeric().withMessage('Amount must be a number').notEmpty().withMessage('Amount is required')
-], createInventoryItem); 
-router.get('/', getInventoryItems);
+], auth, createInventoryItem);
```
Comment (miguelaenlle) [inline backend/src/routes/inventory-item-routes.ts:13]: Looks good to me

Comment (miguelaenlle) [inline backend/src/routes/inventory-item-routes.ts:13]: Inventory Item auth additions look good to me

File: backend/src/controllers/filter-item-controller.ts [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+        let fetchedInventoryItems: any = {};
+        const itemOwner = req.query.itemOwner as string
+        const size = req.query.size as string
+        const name = req.query.name as string
+
+        // Sort
+
+        const sortedBy = req.query.sortedBy as string
+
+        if (itemOwner && size && name) {
```
Comment (miguelaenlle) [inline backend/src/controllers/filter-item-controller.ts]: I have some suggestions on how to simplify the sort logic; let's cover them on the call.

**Review by miguelaenlle — COMMENTED** [review]
Looks good, @AAlchi ! 
- Inventory Item Routes authentication looks good
- Search component looks good

I have some suggestions to simplify the Item sorting logic in `getFilteredInventoryItems`. Can we go over them after tomorrow's call?

Also, create a new branch for each change, so that you have up-to-date code and so I can merge each change without needing the other ones to be ready. I'll explain this in more detail on the call tomorrow.
